[misc] Remove deprecated Python APIs

izackwu opened the pull request:
This PR tries to remove all these deprecated Python APIs and it solves #3590. 

Before merging, we need to make sure all these APIs...

- `@deprecated('kernel_profiler_print()', 'print_kernel_profile_info()')`
- `@deprecated('kernel_profiler_clear()', 'clear_kernel_profile_info()')`
- `@deprecated('memory_profiler_print()', 'print_memory_profile_info()')`
- `@deprecated('ti.cache_shared', 'ti.block_local')`
- `@deprecated('ti.var', 'ti.field')`
- `@deprecated("ti.indices", "ti.axes")`
- `@deprecated('ti.Matrix.transposed(a)', 'a.transpose()')`
- `@deprecated('a.T()', 'a.transpose()')`
- `@deprecated('ti.Matrix.var', 'ti.Matrix.field')`
- `@deprecated('ti.Vector.var', 'ti.Vector.field')`
- `@deprecated('x(i, j)', 'x.get_scalar_field(i, j)')`
- `@deprecated('_bit_struct', 'bit_struct')`
- `@deprecated('_bit_array', 'bit_array')`
- `@deprecated('x.data_type()', 'x.dtype')`
- `@deprecated('x.dim()', 'len(x.shape)')`
- `@deprecated('x.shape()', 'x.shape')`
- `@deprecated('x.get_shape(i)', 'x.shape[i]')`
- `@deprecated('ti.vec(x, y)', 'ti.core_vec(x, y)')`
- `@deprecated('ti.veci(x, y)', 'ti.core_veci(x, y)')`
- `@deprecated('gui.has_key_pressed()', 'gui.get_event()')`

...can be safely deleted by checking other repos under Taichi including

- [x] [taichi-dev/difftaichi](https://github.com/taichi-dev/difftaichi) https://github.com/taichi-dev/difftaichi/pull/53
- [x] [taichi-dev/taichi_elements](https://github.com/taichi-dev/taichi_elements) https://github.com/taichi-dev/taichi_elements/pull/84
- [x] and more 🤔

This PR will follow such steps:

- [x] Remove these APIs from Taichi code
- [x] Make sure all Python tests pass
- [x] Fix other repos' dependence on these outdated APIs

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/taichi-dev/taichi?pullRequest=3725) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/taichi-dev/taichi?pullRequest=3725) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/taichi-dev/taichi?pullRequest=3725) it.</sub>

izackwu commented:
I think I have fixed most things so far. Please take a look @alizhang @k-ye :)

BTW, as for other repos, I'm not sure whether it's really necessary to manually check their dependence on deprecated APIs:

- It's tedious to do so
- They've already locked the version of Taichi in `requirements.txt` like `taichi==0.8.1`, so they won't break
- My personal reason: I'm using MackBook with M1 chip, it's a bit harder to for me set up their various running environments 🥲

So I just fixed these two important ones (as mentioned by @k-ye), while leaving other repos as they are

- https://github.com/taichi-dev/difftaichi/pull/53 
- https://github.com/taichi-dev/taichi_elements/pull/84

izackwu commented on misc/demo_warning.py:
```diff
@@ -1,14 +1,5 @@
-
-
-func()
+ti.dot(x, x)
```
Actually, as I have removed the deprecated APi (`transposed`), this demo will break. So I just picked another deprecated API (yes, there're still some more APis to deprecate in the codebase) for the demo.

ailzhang replied:
Aha I see, makes sense! Thanks!

strongoier replied:
@ailzhang Do we need another issue for other deprecated APIs not covered here?

izackwu replied:
glad to help if we're going to remove all remaining deprecated APIs :)

ailzhang replied:
@izackwu awesome thanks! Would you mind opening an issue as well? Thanks a lot!
